core/src/main/java/org/mule/runtime/core/internal/policy/CompositeOperationPolicy.java

```diff
@@ -248,20 +248,53 @@ public void process(CoreEvent operationEvent,
 
     policySink.next(operationEventForPolicy(operationEvent,
                                             operationExecutionFunction,
-                                            parametersProcessor, callback));
+                                            parametersProcessor,
+                                            new CompositeOperationPolicyExecutorCallback(callback)));
+  }
+
+  private static final class CompositeOperationPolicyExecutorCallback implements ExecutorCallback {
+
+    private final ExecutorCallback callback;
+
+    public CompositeOperationPolicyExecutorCallback(ExecutorCallback callback) {
+      this.callback = callback;
+    }
+
+    @Override
+    public void complete(Object o) {
+      callback.complete(o);
+    }
+
+    @Override
+    public void error(Throwable throwable) {
+      if (throwable instanceof MessagingException) {
+        CoreEvent event = ((MessagingException) throwable).getEvent();
+        OperationPolicyContext context = OperationPolicyContext.from(event);
+        // Restore original event if chain failed before or during operation
+        if (!event.equals(context.getOriginalEvent()) && context.getNextOperationResponse() == null) {
+          throwable = new MessagingException(context.getOriginalEvent(), ((MessagingException) throwable));
+        }
+      }
+
+      callback.error(throwable);
```

Comment:
this should happen in a finally clause. If another bug makes this method fail (let's say, somebody breaks `OperationPolicyContext.from`) , then the callback is never invoked and the process hungs, making it a painful to trace.